README.md
# Wildcat

A Ruby Gem that exposes the wonderful goodness of the [Pro Football API](http://github.com/sjtipton/pro_football_api)

## Installation

Clone the repo from Github

    git clone git@github.com:sjtipton/wildcat.git

Install any dependencies

    bundle install

Build the gem

    bundle exec rake build

Install the gem

    ruby -S gem install ./pkg/wildcat-0.0.8.gem

To vendorize into a project

    cd path/to/project

    gem unpack wildcat -v 0.0.8 --target ./vendor/gems


CHANGELOG.md
## Version 0.0.8

**Add opts hash to Wildcat::Game.all to optionally return games by week**

## Version 0.0.7

**Add instance methods on @game**
* `@game.home_team`
* `@game.away_team`

## Version 0.0.6

**Return a valid Time object for :played_at**

## Version 0.0.5

**Expose @team.games**

## Version 0.0.4

**Use ActiveModel version 3.2.3**

## Version 0.0.3

**Expose Game.find_by_team**
**Use Hydra stub helpers**

## Version 0.0.2

**Expose Team resource**
**Expose Game resource**
**Install RSpec**
**Add Wildcat Configuration**

## Version 0.0.1

**Initial build of Wildcat**


spec/factories.rb
FactoryGirl.define do

  factory :team, class: Wildcat::Team do
    id            { SecureRandom.random_number(1e2.to_i) }
    name          { ["Baltimore Ravens", "Carolina Panthers"].sample }
    nickname      { name.split(' ').last }
    abbreviation  { ["BAL", "CAR"].sample }
    location      { "#{Forgery(:address).city}, #{Forgery(:address).state_abbrev}" }
    conference    { %w(AFC NFC).sample }
    division      { %w(North South East West).sample }
  end

  factory :game, class: Wildcat::Game do
    id            { SecureRandom.random_number(1e3.to_i) }
    label         { "Ravens at Panthers" }
    season        { Forgery(:date).year }
    stadium       { "Bank of America Stadium" }
    week          { (1..17).to_a.sample }
    home_team_id  { SecureRandom.random_number(1e2.to_i) }
    away_team_id  { SecureRandom.random_number(1e2.to_i) }
    played_at     { Time.now }
  end
end


spec/spec_helper.rb
require 'rspec'

require 'rubygems'
require 'bundler/setup'

Bundler.require(:default)

require 'ap'
require 'factory_girl'
require 'forgery'
require 'shoulda-matchers'

# Requires supporting files with custom matchers and macros, etc,
# in ./support/ and its subdirectories.
Dir["#{File.dirname(__FILE__)}/support/**/*.rb"].each {|f| require f}

RSpec.configure do |config|
end

HYDRA = Typhoeus::Hydra.new(max_concurrency: 20) # keep from killing some servers

Wildcat::Config.protocol = "https"
Wildcat::Config.host = "localhost:3000"
Wildcat::Config.api_version = "api/v1"
Wildcat::Config.timeout = 10000
Wildcat::Config.auth_token = "CZxpDEnBSs9ZWLU9XFAz"
Wildcat::Config.hydra ||= HYDRA

FactoryGirl.find_definitions


spec/wildcat_spec.rb
require "spec_helper"

describe Wildcat do

  it "should show the version" do
    Wildcat::VERSION.should_not be_nil
  end
end


lib/wildcat/team.rb
class Wildcat::Team
  include ActiveModel::Validations
  include ActiveModel::Serializers::JSON

  self.include_root_in_json = false

  ATTRIBUTES = [ :id,
                 :name,
                 :nickname, 
                 :abbreviation, 
                 :location, 
                 :conference,
                 :division ]

  attr_accessor *ATTRIBUTES

  validates :name, :nickname, :abbreviation, :location, :conference, :division,
            presence: true

  def initialize(attributes = {})
    self.attributes = attributes
  end

  def attributes
    ATTRIBUTES.inject(ActiveSupport::HashWithIndifferentAccess.new) do |result, key|
      result[key] = read_attribute_for_validation(key)
      result
    end
  end

  def attributes=(attrs)
    attrs.each_pair do |k, v|
      if "#{k}" == "source_id"
        send("#{k}=", v.to_i)
      else
        send("#{k}=", v)
      end
    end
  end

  def read_attribute_for_validation(key)
    send(key)
  end

  def self.find(id)
    request = Typhoeus::Request.new( Wildcat::Config.base_url +
                                      "/teams/#{id}?auth_token=#{Wildcat::Config.auth_token}",
                                      { method: :get,
                                      timeout: Wildcat::Config.timeout,
                                      headers: {:Accept => "application/json", "Content-Type" => "application/json"} })

    request.on_complete do |response|
      if response.code == 200
        parsed = Yajl::Parser.parse(response.body, symbolize_keys: true)
        yield new(id: parsed[:id],
                name: parsed[:name],
            nickname: parsed[:nickname],
        abbreviation: parsed[:abbreviation],
            location: parsed[:location],
          conference: parsed[:conference],
            division: parsed[:division])
      elsif response.code == 404
        parsed = Yajl::Parser.parse(response.body, symbolize_keys: true)
        raise Wildcat::ResourceNotFound.new(*parsed[:errors])
      elsif response.code == 401
        raise Wildcat::UnauthorizedAccess.new
      elsif (500..599).cover? response.code
        raise Wildcat::ServiceUnavailable.new
      else
        raise Wildcat::Error.new(Wildcat::Error::GENERAL_ERROR,
                                { error: response.curl_error_message })
      end
    end

    Wildcat::Config.hydra.queue(request)
  end

  def self.all
    request = Typhoeus::Request.new( Wildcat::Config.base_url +
                                      "/teams?auth_token=#{Wildcat::Config.auth_token}",
                                      { method: :get,
                                      timeout: Wildcat::Config.timeout,
                                      headers: {:Accept => "application/json", "Content-Type" => "application/json"} })

    request.on_complete do |response|
      if response.success?
        teams = []
        parsed = Yajl::Parser.parse(response.body, symbolize_keys: true)
        parsed.each do |result|
          teams << new(id: result[:id],
                     name: result[:name],
                 nickname: result[:nickname],
             abbreviation: result[:abbreviation],
                 location: result[:location],
               conference: result[:conference],
                 division: result[:division])
        end
        yield teams
      elsif response.code == 401
        raise Wildcat::UnauthorizedAccess.new
      elsif (500..599).cover? response.code
        raise Wildcat::ServiceUnavailable.new
      else
        raise Wildcat::Error.new(Wildcat::Error::GENERAL_ERROR,
                                { error: response.curl_error_message })
      end
    end

    Wildcat::Config.hydra.queue(request)
  end

  def games
    Wildcat::Game.find_by_team(self.id) { |g| @games = g }
    Wildcat::Config.hydra.run
    @games
  end
end


lib/wildcat/error.rb
module Wildcat
  class Error < StandardError
    attr_reader :status, :message, :errors

    HTTP_BAD_REQUEST         = 400
    HTTP_UNAUTHORIZED        = 401
    HTTP_RESOURCE_NOT_FOUND  = 404
    HTTP_SERVICE_UNAVAILABLE = 503

    UNAUTHORIZED_ERROR_MESSAGE   =  { error: "Invalid authentication token." }
    SERVICE_DOWN_ERROR_MESSAGE   =  { error: "We're sorry but something went wrong." }
    BAD_URI_ERROR_MESSAGE        =  { error: "Routing error. Please check request uri" }

    BAD_REQUEST         = [HTTP_BAD_REQUEST, "Validation failure"]
    RESOURCE_NOT_FOUND  = [HTTP_RESOURCE_NOT_FOUND, "Resource not found"]
    UNAUTHORIZED        = [HTTP_UNAUTHORIZED, "You need to sign in or sign up before continuing."]
    SERVICE_UNAVAILABLE = [HTTP_SERVICE_UNAVAILABLE, "Service unavailable."]
    GENERAL_ERROR       = [0, "Unknown error."]

    def initialize(info, *args)
      @status, msg = *info
      @message = sprintf(msg, *args)
      @errors = *args
      super(@message)
    end

    def to_json(options = nil)
      Yajl::Encoder.encode({:code => @status, :description => @message, :errors => @errors})
    end
  end

  class UnauthorizedAccess < Error

    def initialize
      super(UNAUTHORIZED, UNAUTHORIZED_ERROR_MESSAGE)
    end
  end

  class ServiceUnavailable < Error

    def initialize
      super(SERVICE_UNAVAILABLE, SERVICE_DOWN_ERROR_MESSAGE)
    end
  end

  class ResourceNotFound < Error

    def initialize(*args)
      super(RESOURCE_NOT_FOUND, *args)
    end
  end
end


lib/wildcat/game.rb
class Wildcat::Game
  include ActiveModel::Validations
  include ActiveModel::Serializers::JSON

  self.include_root_in_json = false

  ATTRIBUTES = [ :id,
                 :away_team_id,
                 :home_team_id, 
                 :label, 
                 :played_at, 
                 :season,
                 :stadium,
                 :week ]

  attr_accessor *ATTRIBUTES

  validates :away_team_id, :home_team_id, :label, :season, :stadium,
            presence: true

  def initialize(attributes = {})
    self.attributes = attributes
  end

  def attributes
    ATTRIBUTES.inject(ActiveSupport::HashWithIndifferentAccess.new) do |result, key|
      result[key] = read_attribute_for_validation(key)
      result
    end
  end

  def attributes=(attrs)
    attrs.each_pair do |k, v|
      if "#{k}" == "source_id"
        send("#{k}=", v.to_i)
      else
        send("#{k}=", v)
      end
    end
  end

  def read_attribute_for_validation(key)
    send(key)
  end

  def self.find(id)
    request = Typhoeus::Request.new( Wildcat::Config.base_url +
                                      "/games/#{id}?auth_token=#{Wildcat::Config.auth_token}",
                                      { method: :get,
                                      timeout: Wildcat::Config.timeout,
                                      headers: {:Accept => "application/json", "Content-Type" => "application/json"} })

    request.on_complete do |response|
      if response.code == 200
        parsed = Yajl::Parser.parse(response.body, symbolize_keys: true)
        yield new(id: parsed[:id],
        away_team_id: parsed[:away_team_id],
        home_team_id: parsed[:home_team_id],
               label: parsed[:label],
           played_at: Time.parse(parsed[:played_at]),
              season: parsed[:season],
             stadium: parsed[:stadium],
                week: parsed[:week])
      elsif response.code == 404
        parsed = Yajl::Parser.parse(response.body, symbolize_keys: true)
        raise Wildcat::ResourceNotFound.new(*parsed[:errors])
      elsif response.code == 401
        raise Wildcat::UnauthorizedAccess.new
      elsif (500..599).cover? response.code
        raise Wildcat::ServiceUnavailable.new
      else
        raise Wildcat::Error.new(Wildcat::Error::GENERAL_ERROR,
                                { error: response.curl_error_message })
      end
    end

    Wildcat::Config.hydra.queue(request)
  end

  def self.all(opts={})
    opts.symbolize_keys!
    params = opts.reject { |k,v| v.blank? }
    query_params = { auth_token: Wildcat::Config.auth_token }.merge(params).to_param
    request = Typhoeus::Request.new( Wildcat::Config.base_url +
                                      "/games?#{query_params}",
                                      { method: :get,
                                      timeout: Wildcat::Config.timeout,
                                      headers: {:Accept => "application/json", "Content-Type" => "application/json"} })

    request.on_complete do |response|
      if response.success?
        games = []
        parsed = Yajl::Parser.parse(response.body, symbolize_keys: true)
        parsed.each do |result|
          games << new(id: result[:id],
             away_team_id: result[:away_team_id],
             home_team_id: result[:home_team_id],
                    label: result[:label],
                played_at: Time.parse(result[:played_at]),
                   season: result[:season],
                  stadium: result[:stadium],
                     week: result[:week])
        end
        yield games
      elsif response.code == 401
        raise Wildcat::UnauthorizedAccess.new
      elsif (500..599).cover? response.code
        raise Wildcat::ServiceUnavailable.new
      else
        raise Wildcat::Error.new(Wildcat::Error::GENERAL_ERROR,
                                { error: response.curl_error_message })
      end
    end

    Wildcat::Config.hydra.queue(request)
  end

  def self.find_by_team(id)
    request = Typhoeus::Request.new( Wildcat::Config.base_url +
                                      "/teams/#{id}/games?auth_token=#{Wildcat::Config.auth_token}",
                                      { method: :get,
                                      timeout: Wildcat::Config.timeout,
                                      headers: {:Accept => "application/json", "Content-Type" => "application/json"} })

    request.on_complete do |response|
      if response.success?
        games = []
        parsed = Yajl::Parser.parse(response.body, symbolize_keys: true)
        parsed.each do |result|
          games << new(id: result[:id],
             away_team_id: result[:away_team_id],
             home_team_id: result[:home_team_id],
                    label: result[:label],
                played_at: Time.parse(result[:played_at]),
                   season: result[:season],
                  stadium: result[:stadium],
                     week: result[:week])
        end
        yield games
      elsif response.code == 401
        raise Wildcat::UnauthorizedAccess.new
      elsif (500..599).cover? response.code
        raise Wildcat::ServiceUnavailable.new
      else
        raise Wildcat::Error.new(Wildcat::Error::GENERAL_ERROR,
                                { error: response.curl_error_message })
      end
    end

    Wildcat::Config.hydra.queue(request)
  end

  def home_team
    Wildcat::Team.find(self.home_team_id) { |team| @team = team }
    Wildcat::Config.hydra.run
    @team
  end

  def away_team
    Wildcat::Team.find(self.away_team_id) { |team| @team = team }
    Wildcat::Config.hydra.run
    @team
  end
end
